src/pareto.rkt

```diff
@@ -17,6 +29,72 @@
   (for/list ([ppt (in-list curve)])
     (struct-copy pareto-point ppt [data (f (pareto-point-data ppt))])))
 
+;; Takes two lists of `pareto-point` structs that are Pareto-optimal
+;; and returns the Pareto-optimal subset of their union.
+(define (pareto-union curve1 curve2)
+  (let loop ([curve1 curve1] [curve2 curve2])
+    ; The curve is sorted so that highest accuracy is first
+    (match* (curve1 curve2)
+      [('() _) curve2]
+      [(_ '()) curve1]
+      [((cons ppt1 rest1) (cons ppt2 rest2))
+       (match (pareto-compare ppt1 ppt2)
+         ['<
+          (loop curve1 rest2)]
+         ['>
+          (loop rest1 curve2)]
+         ['=
+          (define joint-data (append (pareto-point-data ppt1) (pareto-point-data ppt2)))
+          (define joint (struct-copy pareto-point ppt1 [data joint-data]))
+          (cons joint (loop rest1 rest2))]
+         ['<>
+          (if (< (pareto-point-error ppt1) (pareto-point-error ppt2))
+              (cons ppt1 (loop rest1 curve2))
+              (cons ppt2 (loop curve1 rest2)))])])))
+
+;; Takes a pareto frontier and returns the points that form a convex frontier.
+(define (pareto-convex ppts)
+  (let loop ([ppts* '()] [ppts ppts])
+    (match ppts
+     [(list p0 p1 p2 pns ...)
+      (match-define (pareto-point p0x p0y _) p0)
+      (match-define (pareto-point p1x p1y _) p1)
+      (match-define (pareto-point p2x p2y _) p2)
+      ; if { p0, p1, p2 } are not convex:
+      ;   discard p1
+      ;   try backtracking one point (if not continue)
+      ; else move forward one point
+      (define m01 (/ (- p1y p0y) (- p1x p0x)))
+      (define m12 (/ (- p2y p1y) (- p2x p1x)))
+      (match* ((> m12 m01) (null? ppts*))
+        [(#t #t) (loop ppts* (append (list p0 p2) pns))]
+        [(#t #f) (loop (rest ppts*) (append (list (first ppts*) p0 p2) pns))]
+        [(#f _) (loop (cons p0 ppts*) (append (list p1 p2) pns))])]
+     [_ (append (reverse ppts*) ppts)])))
+
+;; Takes a list of `pareto-point` structs
+;; and returns the Pareto-optimal subset.
+(define (pareto-minimize ppts #:convex? [convex? #f])
+  (define ppts* (sort ppts < #:key pareto-point-cost))
+  (define minimized
+    (for/fold ([minimized '()]) ([ppt (in-list ppts*)])
+      (pareto-union (list ppt) minimized)))
+  (if convex? (pareto-convex minimized) minimized))
```

Comment:
Is it necessary to pass `convex` here as a keyword? It seems like it would be simpler just to do that logic in the caller.